okio/src/main/java/okio/BufferCursor.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio;
+
+/**
+ * Direct access to the underlying byte arrays in a buffer.
+ *
+ * Note that cursors are stateful and can be reused.
+ *
+ * It is an error to mutate a buffer that is currently being operated on with a cursor.
+ *
+ * Currently this cursor only supports reading. It is an error to mutate the fields of this cursor
+ * or the values of the data[] array.
+ *
+ * Access should be in acquire/release pairs. This guarded access may be used to detect bugs where
+ * cursors get out of sync with their underlying buffers.
+ *
+ * TODO(jwilson): nice, human-readable docs
+ * TODO(jwilson): support growing the buffer at the end
+ * TODO(jwilson): support acquiring for write (with seek() that gets writable segments)
+ */
+public final class BufferCursor {
+  private Buffer buffer;
+  public long size = -1L;
+
+  private Segment segment;
+  public long offset = -1L;
+  public byte[] data;
+  public int pos = -1;
+  public int limit = -1;
+
+  public void acquireReadOnly(Buffer buffer) {
+    if (this.buffer != null) throw new IllegalStateException("already acquired");
+
+    this.buffer = buffer;
+    this.size = buffer.size;
+  }
+
+  /**
+   * Seeks to the next range of bytes, advancing the offset by {@code limit - pos}. Returns the size
+   * of the readable range (at least 1), or -1 if we have reached the end of the buffer and there
+   * are no more bytes to read.
+   */
+  public int next() {
+    if (offset == size) throw new IllegalStateException();
+    if (offset == -1L) return seek(0L);
+    return seek(offset + (limit - pos));
+  }
+
+  /**
+   * Reposition the cursor so that the data at {@code offset} is readable at {@code data[pos]}.
+   * Returns the number of bytes readable in {@code data} (at least 1), or -1 if we have reached the
+   * end of the buffer and there are no more bytes to read.
+   */
+  public int seek(long offset) {
+    if (offset < 0 || offset > size) {
+      throw new ArrayIndexOutOfBoundsException(String.format("offset=%s > size=%s", offset, size));
+    }
+
+    if (offset == size) {
+      this.segment = null;
+      this.offset = offset;
+      this.data = null;
+      this.pos = -1;
+      this.limit = -1;
+      return -1;
+    }
+
+    // Navigate to the segment that contains `offset` starting from our current segment if possible.
+    Segment next;
+    long nextOffset;
+    if (segment == null || this.offset > offset) {
```

Comment:
Is seeking backwards going to be useful?  Or should the cursor just be released and re-acquired?  I'm not immediately seeing the usefulness but I'm guessing you have ideas?

Reply:
Just for binary searching a buffer.